### PR TITLE
Add suggest_parentheses rule for Verible::Linter

### LINKED ISSUE #383: Recommend parentheses around certain expression patterns
Though parsers and compilers know fully well the precedence of operators, humans can still benefit from extra parenthesis.

example:
https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#parentheses

Task: detect such situations (starting with the above ternary expression example) and report them

Suggest: subclassing a SyntaxTreeLintRule and for various expression types, looking downwards at its operands and comparing operators/precedences.  

Feel free to suggest more specific operator classes and cases.

### PR DESCRIPTION
**Pull request by omarahmed1111**
This PR introduces:
* Implementation of suggest_parentheses rule for the linter.
* Change kTernaryExpression to kConditionExpression globally.

suggest-parentheses rule is a new rule for the linter to suggest parentheses around nested expressions. The rule is not enabled by-default. Now, the rule only suggest parentheses around nested kConditionExpression in kConditionExpression.

Fixes #383

### CODE REVIEW

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule_test.cc:35]: As per the earlier comment, this should also be accepted with un-parenthesized `condition_c? c : d`.  (Can you add this as a separate test case?)

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule_test.cc:37]: Add some test cases that expect to find violations like this:

```c++
constexpr int kTag = 1;  // don't care
const std::initializer_list<LintTestCase> kSuggestParenthesesTestCases = {
  ...
  {"module m;\nassign foo = condition_a? ", {kTag, "condition_b? a : b"}, " : c;\nendmodule"},
};
```

This notation `{kTag, "condition_b? a : b"}` says: "expect to find a violation on this exact substring range."

File: verilog/analysis/checkers/suggest_parentheses_rule.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule.cc:68]: Instead of highlighting the whole `node`, the violation's text should narrow down to `*trueCase` because that is where parentheses can be added to remedy the problem.

File: verilog/analysis/checkers/suggest_parentheses_rule.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule.cc:38]: I would make this message more specific to the implemented check:

"Parenthesize condition expressions that appear in the true-clause of another condition expression."

As more checks are added, they will get their own specific messages.

File: verilog/analysis/checkers/suggest_parentheses_rule.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule.cc:43]: This message can be generic: "Recommend extra parentheses around subexpressions where it helps readability."

File: verilog/analysis/checkers/suggest_parentheses_rule.h
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule.h:24]: // Suggests where extra parentheses can help readability.
// For example:
//   a1 ? a2 ? b : c : d
// would be improved as the equivalent:
//   a1 ? (a2 ? b : c) : d

File: verilog/analysis/checkers/suggest_parentheses_rule.h
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule.h:43]: // Diagnostic message for rule violations.

(nothing to do with constraints)

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule_test.cc:38]: Suggest testing a few more contexts like:

* `parameter foo = expresssion;`
*  
```systemverilog
always @(posedge clk) begin
  left <= expression;
end
```
* 
```systemverilog
function f;
  g = h(expression);
endfunction
```

File: verilog/parser/verilog.y
Comment (fangism) [inline verilog/parser/verilog.y:383]: Revert, syntax change isn't needed.

File: verilog/parser/verilog.lex
Comment (fangism) [inline verilog/parser/verilog.lex:557]: Revert, lexical change isn't needed.

**Review by fangism — CHANGES_REQUESTED** [review]

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (omarahmed1111) [reply verilog/analysis/checkers/suggest_parentheses_rule_test.cc:35]: I didn't notice this part in this guide. Okay, I will modify that and provide more test cases on this part. :)

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (omarahmed1111) [reply verilog/analysis/checkers/suggest_parentheses_rule_test.cc:37]: Truth to be said, I was going to ask about this part and how to do that. Thanks for this info. :)

Comment (fangism) [reply verilog/analysis/checkers/suggest_parentheses_rule_test.cc:37]: Glad that helped!  Don't forget to "request review" when ready.  Thanks for working on this again.

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (fangism) [reply verilog/analysis/checkers/suggest_parentheses_rule_test.cc:38]: Also suggest a few test cases where the true-clause is a different expression, like a shift operation or addition, where no parentheses are recommended (yet).

File: verilog/analysis/checkers/suggest_parentheses_rule_test.cc
Comment (fangism) [inline verilog/analysis/checkers/suggest_parentheses_rule_test.cc:28]: Since this contains a mix of positive and negative tests, rename `AcceptTests` to something like `Various`.

File: verilog/CST/expression.cc
Comment (fangism) [inline verilog/CST/expression.cc:78]: This will terminate with an assert failure if passed a leaf, so suggest checking `expr.Kind()` first before doing this cast.

File: verilog/CST/expression.cc
Comment (fangism) [inline verilog/CST/expression.cc:99]: Let's add unit tests for these, so that should any of these get restructured, we'll know to fix them in one place.

[Example unit tests](https://cs.opensource.google/verible/verible/+/master:verilog/CST/declaration_test.cc).
Or look at many other CST/*_test.cc files.

A typical recipe for unit testing these extractor functions:
1) FindAll of the outer structure of interest (might need to write a quick function for this too), in this case, condition expressions.
2) Write test cases, tagging the substring of text that corresponds to what an accessor function should return: condition, true case, false case.
3) Pass the array of test cases to `TestVerilogSyntaxRangeMatches` (this is where you call FindAll and possibly other functions to refine your search).

**Review by fangism — CHANGES_REQUESTED** [review]
Thanks again for another round, we're almost done.

**Review by fangism — APPROVED** [review]
Nice job on the unit tests!  Thank you again for this contribution, others will be able to follow what you've done.

File: .vscode/settings.json
Comment (omarahmed1111) [reply .vscode/settings.json:1]: Okay sry, I added them by mistake

**Comment by omarahmed1111** [comment]
> Nice job on the unit tests! Thank you again for this contribution, others will be able to follow what you've done.

Thank you for the great review and help. Glad for my first contribution here :)

File: .vscode/settings.json
Comment (omarahmed1111) [reply .vscode/settings.json:1]: Do I need to sumbit another update with them deleted or they will be filtered out automatically on merge ?

Comment (fangism) [reply .vscode/settings.json:1]: Your work is done, we took care of it.  (Just waiting for internal approval, but expecting delays over the holiday season.)

Comment (omarahmed1111) [reply .vscode/settings.json:1]: Got it, Thanks

**Pull request closed without merging**
